### PR TITLE
Fix skipping properties with `undefined` value in SDK query parsing

### PR DESCRIPTION
**Pull request by br41nslug**
## Scope

What's changed:

- Fixed the error case of `search: undefined` actually searching for the string `'undefined'` (linked issue)
- Fixed `undefined` handling for the other query parameters
- Attempted to make the query formatting conditions more explicit
- Added tests for the `formatFields` and `queryToParams` utilities
- Fixed the `aggregate` type by adding `Partial` (the tests showed that all aggregation keys were required before)
- Fixed the `aggregate` type by adding `& Record<string, unknown>` allowing for custom properties without type errors
- Moved the `request.test.ts` file next to the `request.ts` file (to consistently keep functional test files next to the file they're testing and type tests in the `test/` folder)

## Potential Risks / Drawbacks

- Something could break if relying on the old fall-through into custom parameter logic for "knownQueryKeys". I covered the edge cases I could think of with explicit logic to handle them.

## Tested Scenarios

- See the added test file

## Review Notes / Questions

- Did I miss any edge cases in the tests?

## Checklist

- [x] Added or updated tests

---

Fixes #25930

### CODE REVIEW

File: sdk/src/rest/utils/query-to-params.ts [hunk 1]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
-		// covers empty array and undefined
-		params['sort'] = typeof query.sort === 'string' ? query.sort : query.sort.join(',');
+	if (query.sort) {
+		if (Array.isArray(query.sort) && query.sort.length > 0) {
+			params['sort'] = query.sort.join(',');
+		}
+
+		if (typeof query.sort === 'string' && query.sort.length > 0) {
+			params['sort'] = query.sort;
+		}
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]:
```suggestion
	if (query.sort && query.sort.length > 0) {
		if (Array.isArray(query.sort)) {
			params['sort'] = query.sort.join(',');
		}

		if (typeof query.sort === 'string') {
			params['sort'] = query.sort;
		}
```

Comment (ComfortablyCoding) [reply sdk/src/rest/utils/query-to-params.ts]: Alternatively
```suggestion
	if (query.sort) {
		if (Array.isArray(query.sort) && query.sort.length > 0) {
			params['sort'] = query.sort.join(',');
		}

		if (typeof query.sort === 'string') {
			params['sort'] = query.sort;
		}
```

File: sdk/src/rest/utils/query-to-params.ts [hunk 2]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
 	}
 
-	if (typeof query.limit === 'number' && query.limit >= -1) {
-		params['limit'] = String(query.limit);
+	if ('limit' in query) {
+		if (typeof query.limit === 'number' && query.limit >= -1) {
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]: I wonder if we should be ignoring `-1`? The API errors if an invalid limit is provided which would provide more meaningful feedback than ignoring.

File: sdk/src/rest/utils/query-to-params.ts [hunk 3]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
 
-	if (query.groupBy && query.groupBy.length > 0) {
-		params['groupBy'] = query.groupBy.join(',');
+	if (query.groupBy) {
+		if (Array.isArray(query.groupBy) && query.groupBy.length > 0) {
+			params['groupBy'] = query.groupBy.join(',');
+		}
+
+		// backwards JS compatibility for string pages
+		if (typeof query.groupBy === 'string' && String(query.groupBy).length > 0) {
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]: An empty string is falsey, this check should be unnecessary.
```suggestion
		if (typeof query.groupBy === 'string') {
```

File: sdk/src/rest/utils/query-to-params.ts [hunk 4]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
 	return fields.flatMap((value) => walkFields(value));
 };
 
+const knownQueryKeys = [
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts:46]: Is it expected for the following keys to be missing?
- `version`
- `versionRaw`
- `export`

File: sdk/src/rest/utils/query-to-params.ts [hunk 5]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
+
+		// backwards JS compatibility for string pages
+		if (typeof query.groupBy === 'string' && String(query.groupBy).length > 0) {
+			params['groupBy'] = query.groupBy;
+		}
 	}
 
 	for (const [key, value] of Object.entries(query)) {
-		if (key in params) continue;
+		if (key in params || knownQueryKeys.includes(key)) continue;
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]: As params will only ever contain knownQueryKeys we can remove the `key in params` check
 ```suggestion
		if (knownQueryKeys.includes(key)) continue;
```

File: sdk/src/rest/utils/query-to-params.ts [hunk 6]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
 	}
 
-	if (query.filter && Object.keys(query.filter).length > 0) {
+	if (query.filter && typeof query.filter === 'object' && Object.keys(query.filter).length > 0) {
 		params['filter'] = JSON.stringify(query.filter);
 	}
 
-	if (query.search) {
-		// covers both empty string and undefined
+	if (query.search && typeof query.search === 'string' && query.search.length > 0) {
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]:
```suggestion
	if (query.search && typeof query.search === 'string') {
```

File: sdk/src/rest/utils/query-to-params.ts [hunk 7]
```diff
@@ -41,64 +41,120 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
 	}
 
-	if (query.groupBy && query.groupBy.length > 0) {
-		params['groupBy'] = query.groupBy.join(',');
+	if (query.groupBy) {
+		if (Array.isArray(query.groupBy) && query.groupBy.length > 0) {
+			params['groupBy'] = query.groupBy.join(',');
+		}
+
+		// backwards JS compatibility for string pages
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts]: Small nitpick 😅 
```suggestion
		// backwards JS compatibility for `groupBy: "id,name"`
```

File: sdk/src/rest/utils/query-to-params.test.ts [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+		});
+
+		test('should accept a direct string (type workaround)', () => {
+			const result = queryToParams({
+				// @ts-ignore
+				fields: 'id,name',
+			});
+
+			expect(result).toEqual({ fields: 'id,name' });
+		});
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.test.ts:139]: For consistency we should add a test for `fields: undefined`

File: sdk/src/rest/utils/query-to-params.ts (hunk 4 above)
Comment (br41nslug) [reply sdk/src/rest/utils/query-to-params.ts:46]: I skipped doing the `export` one because the SDK technically cannot download the results as file

File: sdk/src/rest/utils/query-to-params.ts (hunk 2 above)
Comment (br41nslug) [reply sdk/src/rest/utils/query-to-params.ts]: Happy to discuss that validation further but since that was already the case it's out of the scope of this PR

File: sdk/src/rest/utils/query-to-params.ts [hunk 9]
```diff
@@ -41,69 +43,114 @@ export const formatFields = (fields: (string | Record<string, any>)[]) => {
...
+/**
+ * Local utility functions
+ */
+const isBoolean = (value: unknown): value is boolean => typeof value === 'boolean';
+const isString = (value: unknown): value is string => typeof value === 'string' && Boolean(value);
+const isNumber = (value: unknown): value is number => typeof value === 'number';
+const isArray = (value: unknown): value is unknown[] => Array.isArray(value) && value.length > 0;
+
+const isObject = (value: unknown): value is Record<string, unknown> =>
+	typeof value === 'object' && value !== null && !isArray(value) && Object.keys(value).length > 0;
```
Comment (ComfortablyCoding) [inline sdk/src/rest/utils/query-to-params.ts:72]: As these might be useful elsewhere in the future may be worth to move these to a `helpers.ts` file under `utils`?

Comment (br41nslug) [reply sdk/src/rest/utils/query-to-params.ts:72]: I would say: in the future yes 😄 Currently the naming could be confusing for future re-use like `isArray` returning false for an empty array and same for string. So once we start actually re-using these in the SDK i'd be happy to abstract that to their own file

**Review by ComfortablyCoding — CHANGES_REQUESTED** [review]
Tests for `version` and `versionRaw` are missing but otherwise LGTM 🚀

File: sdk/src/rest/utils/query-to-params.test.ts [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+		const result = queryToParams({
+			version: 'draft',
+		});
+
+		expect(result).toEqual({
+			version: 'draft',
+		});
+	});
+
+	test('should format version parameter', () => {
```
Comment (br41nslug) [inline sdk/src/rest/utils/query-to-params.test.ts]:
```suggestion
	test('should format versionRaw parameter', () => {
```

**Review by ComfortablyCoding — APPROVED** [review]
